src/lib/y2partitioner/format_mount/base.rb

```diff
@@ -62,20 +65,137 @@ def apply_format_options!
         true
       end
 
+      # Performs selected mount options over the device
       def apply_mount_options!
-        return false unless @partition.filesystem
-
-        if @options.mount
-          @partition.filesystem.mount_point = @options.mount_point
-          @partition.filesystem.mount_by = @options.mount_by
-          @partition.filesystem.label = @options.label if @options.label
-          @partition.filesystem.fstab_options = @options.fstab_options
-        else
-          @partition.filesystem.mount_point = ""
-        end
+        return false if filesystem.nil?
+
+        set_mount_point
+        set_mount_options
 
         true
       end
+
+    private
+
+      attr_reader :partition
+
+      attr_reader :options
+
+      def set_mount_point
+        return unless change_mount_point?
+
+        subvolume_actions_before_set_mount_point
+        filesystem.mount_point = mount? ? options.mount_point : ""
+        subvolume_actions_after_set_mount_point
+      end
+
+      def set_mount_options
+        return unless mount?
+
+        filesystem.mount_by = options.mount_by
+        filesystem.label = options.label if options.label
+        filesystem.fstab_options = options.fstab_options
+      end
+
+      # Actions to perform before setting a new mount point
+      #
+      # When the filesystem is btrfs, the not probed subvolumes are deleted.
+      def subvolume_actions_before_set_mount_point
+        delete_new_subvolumes if btrfs?
+      end
+
+      # Actions to perform after setting a new mount point
+      #
+      # When the filesystem is btrfs and root, default proposed subvolumes are added
+      # in case they are not been probed.
+      # When the filesystem is btrfs, the mount point of the resulting subvolumes is updated.
+      # Shadowing control of root btrfs subvolumes is always performed.
+      def subvolume_actions_after_set_mount_point
+        add_proposed_subvolumes if btrfs_root?
+        update_mount_points if btrfs?
+        Y2Storage::Filesystems::Btrfs.refresh_root_subvolumes_shadowing(devicegraph)
+      end
+
+      def delete_new_subvolumes
+        new_subvolumes.each { |s| filesystem.delete_btrfs_subvolume(devicegraph, s.path) }
+      end
+
+      # Subvolumes that have not been probed
+      # @return [Array<Y2Storage::BtrfsSubvolume>]
+      def new_subvolumes
+        devicegraph = DeviceGraphs.instance.system
+        filesystem.btrfs_subvolumes.select { |s| !s.exists_in_devicegraph?(devicegraph) }
+      end
+
+      # Only proposed subvolumes that have not been already probed are added
+      def add_proposed_subvolumes
+        specs = Y2Storage::SubvolSpecification.for_current_product
```

Comment:
I really, really feel we should use this PBI to kill `Planned::BtrfsSubvolume` I see a lot of duplicity (in code and responsibilities introduced in the love triangle `BtrfsSubvolume`, `Planned::BtrfsSubvolume`, `SubvolumeSpecification`).